From a5920af765211f4b6426b4092698b80e0b5098ff Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Sun, 4 Oct 2026 19:02:48 +0200
Subject: [PATCH] Fix hyperlinks in body

We are reporting `body-hyperlinks` capability, but it did not work
because hyperlinks are not allowed in pango markup. Fix this by
removing links from markup before using `pango_parse_markup` to
validate markup.
---
 src/nd-bubble.c           |   2 +-
 src/nd-notification-box.c |   2 +-
 src/nd-notification.c     | 150 ++++++++++++++++++++++++++++++++++++++
 src/nd-notification.h     |   2 +
 4 files changed, 154 insertions(+), 2 deletions(-)

diff --git a/src/nd-bubble.c b/src/nd-bubble.c
index caeeab4..4e737d1 100644
--- a/src/nd-bubble.c
+++ b/src/nd-bubble.c
@@ -631,7 +631,7 @@ set_notification_text (NdBubble   *bubble,
         g_free (str);
         gtk_widget_show_all (GTK_WIDGET (bubble));
 
-        if (pango_parse_markup (body, -1, 0, NULL, NULL, NULL, NULL))
+        if (validate_markup (body))
                 gtk_label_set_markup (GTK_LABEL (bubble->priv->body_label), body);
         else {
                 gchar *tmp;
diff --git a/src/nd-notification-box.c b/src/nd-notification-box.c
index 8e94180..35d3f45 100644
--- a/src/nd-notification-box.c
+++ b/src/nd-notification-box.c
@@ -222,7 +222,7 @@ update_notification_box (NdNotificationBox *notification_box)
 
         /* body */
         body = nd_notification_get_body (notification_box->priv->notification);
-        if (pango_parse_markup (body, -1, 0, NULL, NULL, NULL, NULL))
+        if (validate_markup (body))
                 gtk_label_set_markup (GTK_LABEL (notification_box->priv->body_label), body);
         else {
                 gchar *tmp;
diff --git a/src/nd-notification.c b/src/nd-notification.c
index cc20309..575df4d 100644
--- a/src/nd-notification.c
+++ b/src/nd-notification.c
@@ -77,6 +77,124 @@ get_next_notification_serial (void)
         return serial;
 }
 
+static void
+start_element_cb (GMarkupParseContext  *context,
+                  const gchar          *element_name,
+                  const gchar         **attribute_names,
+                  const gchar         **attribute_values,
+                  gpointer              user_data,
+                  GError              **error)
+{
+  GString *str;
+  gint i;
+
+  if (g_strcmp0 (element_name, "a") == 0)
+    return;
+
+  str = user_data;
+
+  g_string_append (str, "<");
+  g_string_append (str, element_name);
+
+  for (i = 0; attribute_names[i] != NULL; i++)
+    {
+      gchar *tmp;
+
+      tmp = g_markup_escape_text (attribute_values[i], -1);
+      g_string_append_printf (str, " %s=\"%s\"", attribute_names[i], tmp);
+      g_free (tmp);
+    }
+
+  g_string_append (str, ">");
+}
+
+static void
+end_element_cb (GMarkupParseContext  *context,
+                const gchar          *element_name,
+                gpointer              user_data,
+                GError              **error)
+{
+  GString *str;
+
+  if (g_strcmp0 (element_name, "a") == 0)
+    return;
+
+  str = user_data;
+
+  g_string_append_printf (str, "</%s>", element_name);
+}
+
+static void
+text_cb (GMarkupParseContext  *context,
+         const gchar          *text,
+         gsize                 text_len,
+         gpointer              user_data,
+         GError              **error)
+{
+  GString *str;
+  gchar *tmp;
+
+  str = user_data;
+
+  tmp = g_markup_escape_text (text, text_len);
+  g_string_append (str, tmp);
+  g_free (tmp);
+}
+
+static gboolean
+parse_markup (const gchar  *text,
+              gchar       **parsed_markup,
+              GError      **error)
+{
+  GString *str;
+  GMarkupParseContext *context;
+
+  str = g_string_new (NULL);
+  context = g_markup_parse_context_new (&(GMarkupParser) {
+                                          start_element_cb,
+                                          end_element_cb,
+                                          text_cb
+                                        },
+                                        0, str, NULL);
+
+  if (!g_markup_parse_context_parse (context, "<markup>", -1, error))
+    {
+      g_markup_parse_context_free (context);
+      g_string_free (str, TRUE);
+
+      return FALSE;
+    }
+
+  if (!g_markup_parse_context_parse (context, text, -1, error))
+    {
+      g_markup_parse_context_free (context);
+      g_string_free (str, TRUE);
+
+      return FALSE;
+    }
+
+  if (!g_markup_parse_context_parse (context, "</markup>", -1, error))
+    {
+      g_markup_parse_context_free (context);
+      g_string_free (str, TRUE);
+
+      return FALSE;
+    }
+
+  if (!g_markup_parse_context_end_parse (context, error))
+    {
+      g_markup_parse_context_free (context);
+      g_string_free (str, TRUE);
+
+      return FALSE;
+    }
+
+  *parsed_markup = g_string_free (str, FALSE);
+  g_markup_parse_context_free (context);
+
+  return TRUE;
+}
+
 static void
 nd_notification_class_init (NdNotificationClass *class)
 {
@@ -570,3 +688,35 @@ nd_notification_new (const char *sender)
 
         return notification;
 }
+
+gboolean
+validate_markup (const gchar *markup)
+{
+  gchar *parsed_markup;
+  GError *error;
+
+  parsed_markup = NULL;
+  error = NULL;
+
+  if (!parse_markup (markup, &parsed_markup, &error))
+    {
+      g_warning ("%s", error->message);
+      g_error_free (error);
+
+      return FALSE;
+    }
+
+  if (!pango_parse_markup (parsed_markup, -1, 0, NULL, NULL, NULL, &error))
+    {
+      g_warning ("%s", error->message);
+      g_error_free (error);
+
+      g_free (parsed_markup);
+
+      return FALSE;
+    }
+
+  g_free (parsed_markup);
+
+  return TRUE;
+}
diff --git a/src/nd-notification.h b/src/nd-notification.h
index ac96f27..a75027c 100644
--- a/src/nd-notification.h
+++ b/src/nd-notification.h
@@ -87,6 +87,8 @@ void                  nd_notification_action_invoked      (NdNotification *notif
 void                  nd_notification_url_clicked         (NdNotification *notification,
                                                            const char     *url);
 
+gboolean              validate_markup                     (const gchar    *markup);
+
 G_END_DECLS
 
 #endif
-- 
2.55.0

